### PR TITLE
Add implementation for UnitConverter and its related classes

### PR DESCRIPTION
**Pull request by adellibovi**
Hi everyone,

This is an implementation for UnitConverter and all its related classes.
I would love to receive your feedbacks.

Should I split in several commits for code reviewing?

Thanks,

### CODE REVIEW

File: Foundation/Unit.swift [hunk 1]
```diff
@@ -20,32 +20,63 @@ public class UnitConverter : NSObject {
...
     
-    public init(coefficient: Double, constant: Double) { NSUnimplemented() }
+    public init(coefficient: Double, constant: Double) {
+        self.coefficient = coefficient
+        self.constant = constant
+    }
+    
+    public required convenience init?(coder aDecoder: NSCoder) {
+        if aDecoder.allowsKeyedCoding {
+            let coefficient = aDecoder.decodeDouble(forKey: "coefficient")
```
Comment (parkera) [inline Foundation/Unit.swift]: the key here is actually `NS.coefficient`

File: Foundation/Unit.swift [hunk 2]
```diff
@@ -20,32 +20,63 @@ public class UnitConverter : NSObject {
...
+            let constant = aDecoder.decodeDouble(forKey: "constant")
```
Comment (parkera) [inline Foundation/Unit.swift]: and `NS.constant`

File: Foundation/Unit.swift [hunk 3]
```diff
@@ -20,32 +20,63 @@ public class UnitConverter : NSObject {
...
 }
 
 public class UnitConverterLinear : UnitConverter, NSSecureCoding {
     
     
-    public var coefficient: Double { get { NSUnimplemented() } }
+    public private(set) var coefficient: Double
     
-    public var constant: Double { get { NSUnimplemented() } }
+    public private(set) var constant: Double
```
Comment (parkera) [inline Foundation/Unit.swift]: we also need to override `baseUnitValue(fromValue:)` and `value(fromBaseUnitValue:)`. Here is the ObjC implementation:

```objc
- (double)baseUnitValueFromValue:(double)value {
    return (value  * _coefficient) + _constant;
}

- (double)valueFromBaseUnitValue:(double)baseUnitValue {
    return (baseUnitValue + (-1 * _constant)) / _coefficient;
}
```

File: Foundation/Unit.swift [hunk 4]
```diff
@@ -67,62 +98,189 @@ public class Unit : NSObject, NSCopying, NSSecureCoding {
...
+        } else {
+            guard let symbol = aDecoder.decodeObject() as? String
+                else { return nil }
+            self.symbol = symbol
+        }
+    }
+    
+    public func encode(with aCoder: NSCoder) {
+        if aCoder.allowsKeyedCoding {
+            aCoder.encode(self.symbol.bridge(), forKey:"symbol")
```
Comment (parkera) [inline Foundation/Unit.swift]: `NS.symbol`

File: Foundation/Unit.swift [hunk 5]
```diff
@@ -67,62 +98,189 @@ public class Unit : NSObject, NSCopying, NSSecureCoding {
...
+        super.init(symbol: symbol)
+    }
     
     /*
      This class method returns an instance of the dimension class that represents the base unit of that dimension.
      e.g.
      NSUnitSpeed *metersPerSecond = [NSUnitSpeed baseUnit];
      */
-    public class func baseUnit() -> Self { NSUnimplemented() }
+    public class func baseUnit() -> Self { fatalError() }
```
Comment (parkera) [inline Foundation/Unit.swift]: Let's put a message in the fatal error so people know what they did wrong. Here is the one from Darwin:

`*** You must override baseUnit in your class to define its base unit.`

File: Foundation/Unit.swift [hunk 6]
```diff
@@ -332,14 +986,48 @@ public class UnitFuelEfficiency : Dimension {
...
+    
+    public class var litersPer100Kilometers: UnitFuelEfficiency {
+        get {
+            return UnitFuelEfficiency(symbol: Symbol.litersPer100Kilometers, coefficient: Coefficient.litersPer100Kilometers)
+        }
+    }
+    
+    public class var milesPerImperialGallon: UnitFuelEfficiency {
+        get {
+            return UnitFuelEfficiency(symbol: Symbol.milesPerImperialGallon, coefficient: Coefficient.milesPerImperialGallon)
```
Comment (parkera) [inline Foundation/Unit.swift]: This is actually not Linear, but UnitConverterReciprocal, which we didn't expose as API but need to do the correct implementation.

```objc
@interface NSUnitConverterReciprocal : NSUnitConverter <NSSecureCoding>
@property (readonly) double reciprocalValue;
- (instancetype)initWithReciprocalValue:(double)reciprocalValue;
@end
```

```objc
- (double)baseUnitValueFromValue:(double)value {
    return _reciprocalValue / value;
}

- (double)valueFromBaseUnitValue:(double)baseUnitValue {
    return _reciprocalValue / baseUnitValue;
}
```

and here:

```
        NSUnitConverterReciprocal *conv = [[[NSUnitConverterReciprocal alloc] initWithReciprocalValue:282.481] autorelease];
```

File: Foundation/Unit.swift [hunk 7]
```diff
@@ -332,14 +986,48 @@ public class UnitFuelEfficiency : Dimension {
...
+        }
+    }
     
-    public required init?(coder aDecoder: NSCoder) { NSUnimplemented() }
-    public override func encode(with aCoder: NSCoder) { NSUnimplemented() }
+    public class var milesPerGallon: UnitFuelEfficiency {
+        get {
+            return UnitFuelEfficiency(symbol: Symbol.milesPerGallon, coefficient: Coefficient.milesPerGallon)
```
Comment (parkera) [inline Foundation/Unit.swift]: and here:

```objc
        NSUnitConverterReciprocal *conv = [[[NSUnitConverterReciprocal alloc] initWithReciprocalValue:235.215] autorelease];
```

**Comment by parkera** [comment]
This looks pretty great.

In addition to the few comments I had, could we also get at least the basic infrastructure for some unit tests, and some verification that converting units works?

File: Foundation/Unit.swift (hunk 3 above)
Comment (adellibovi) [inline Foundation/Unit.swift]: My bad, forgot to push this one.__

**Comment by adellibovi** [comment]
Hi @parkera thanks for the comments :)
I'm going to make the changes together with the unit tests today.

File: Foundation/Unit.swift (hunk 6 above)
Comment (adellibovi) [inline Foundation/Unit.swift]: @parkera shouldn't be like this?

`- (double)valueFromBaseUnitValue:(double)baseUnitValue {
    return baseUnitValue * _reciprocalValue;
}`

**Comment by adellibovi** [comment]
@parkera just pushed a new version with the changes and unit tests. ✌️

The test covers:
- the linearity of the linear converter
- the bijectivity of every converter
- the corresponding base unit for each unit

Any other case that must be covered?
Thanks

**Comment by parkera** [comment]
Awesome

**Pull request merged**